### PR TITLE
[core][autoscaler][v2] Reconciler - scaling clusters update [10/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR implements the reconcilation updates needed to scale up/down the clusters by consulting the scheduler. 

Also depends on https://github.com/ray-project/ray/pull/42525

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 1]
```diff
@@ -956,6 +988,16 @@ def _handle_stuck_instances(
                     timeout_s=reconcile_config.terminating_status_timeout_s,
                 ),
             ),
+            # If we tried to stop ray on the instance, but it doesn't stop after a long
+            # time, we will transition it back to RAY_RUNNING as the stop/drain somehow
+            # failed. If it had succeed, we should have transition it to RAY_STOPPING
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:993]:
```suggestion
            # failed. If it had succeed, we should have transitioned it to RAY_STOPPING
```

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 2]
```diff
@@ -1021,6 +1063,120 @@ def _warn_stuck_instances(
...
+                    termination_request=terminate_request,
+                )
+            else:
+                # If we would just terminate the cloud instance.
+                updates[terminate_request.instance_id] = IMInstanceUpdateEvent(
+                    instance_id=instance_id,
+                    new_instance_status=IMInstance.TERMINATING,
+                )
+            logger.info(
+                "Terminating {} with {}".format(
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:1156]: This log only applies to the else branch?

File: python/ray/autoscaler/v2/scheduler.py [hunk 3]
```diff
@@ -123,9 +127,112 @@ class SchedulingNode:
...
+
+        # These are the statuses where there's a running ray node or
+        # could eventually run ray.
+        if instance.im_instance.status in (
+            Instance.QUEUED,
+            Instance.REQUESTED,
+            Instance.ALLOCATED,
+            Instance.RAY_INSTALLING,
+            Instance.RAY_RUNNING,
+            Instance.RAY_STOP_REQUESTED,
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:228]: we should be able to derive all states that can be transitioned into RAY_RUNNING

File: src/ray/protobuf/experimental/instance_manager.proto [hunk 4]
```diff
@@ -68,22 +68,24 @@ message Instance {
     // Ray started and connected to gcs - follows from the RAY_INSTALLING state or
     // ALLOCATED.
     RAY_RUNNING = 5;
+    // Ray node should be stopped - follows from the RAY_RUNNING state.
+    RAY_STOP_REQUESTED = 6;
     // Ray node being stopped - follows from RAY_RUNNING.
```
Comment (jjyao) [inline src/ray/protobuf/experimental/instance_manager.proto:73]: needs update

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 5]
```diff
@@ -1021,6 +1063,120 @@ def _warn_stuck_instances(
...
+        to_terminate = reply.to_terminate
+        updates = {}
+        # Add terminating instances.
+        for terminate_request in to_terminate:
+            instance_id = terminate_request.instance_id
+            if autoscaling_config.need_ray_stop():
+                # If we would need to stop/drain ray.
+                updates[terminate_request.instance_id] = IMInstanceUpdateEvent(
+                    instance_id=instance_id,
+                    new_instance_status=IMInstance.RAY_STOP_REQUESTED,
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:1146]: where is the transition code from RAY_STOP_REQUESTED to RAY_STOPPING?

**Review by jjyao — APPROVED** [review]

File: python/ray/autoscaler/v2/instance_manager/reconciler.py (hunk 5 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:1146]: Handles by `_handle_ray_status_transition`

**Pull request merged**